drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1322 @@
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	strncpy(data->hwinfo, argv[1], sizeof(data->hwinfo) - 1);
+}
+
+static bool modem_cellular_is_registered(struct modem_cellular_data *data)
+{
+	return (data->registration_status_gsm == 1)
+		|| (data->registration_status_gsm == 5)
+		|| (data->registration_status_gprs == 1)
+		|| (data->registration_status_gprs == 5)
+		|| (data->registration_status_lte == 1)
+		|| (data->registration_status_lte == 5);
+}
+
+static void modem_cellular_chat_on_cxreg(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+	uint8_t registration_status;
+	bool is_registered;
+
+	is_registered = modem_cellular_is_registered(data);
+
+	if (argc == 2) {
+		registration_status = atoi(argv[1]);
+	} else if (argc == 3) {
+		registration_status = atoi(argv[2]);
+	} else {
+		return;
+	}
+
+	if (strcmp(argv[0], "+CREG: ") == 0) {
+		data->registration_status_gsm = registration_status;
+	} else if (strcmp(argv[0], "+CGREG: ") == 0) {
+		data->registration_status_gprs = registration_status;
+	} else {
+		data->registration_status_lte = registration_status;
+	}
+
+	if (modem_cellular_is_registered(data) == is_registered) {
+		return;
+	}
+
+	if (modem_cellular_is_registered(data)) {
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_REGISTERED);
+	} else {
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_DEREGISTERED);
+	}
+}
+
+MODEM_CHAT_MATCH_DEFINE(ok_match, "OK", "", NULL);
+MODEM_CHAT_MATCHES_DEFINE(allow_match,
+			  MODEM_CHAT_MATCH("OK", "", NULL),
+			  MODEM_CHAT_MATCH("ERROR", "", NULL));
+
+MODEM_CHAT_MATCH_DEFINE(imei_match, "", "", modem_cellular_chat_on_imei);
+MODEM_CHAT_MATCH_DEFINE(cgmm_match, "", "", modem_cellular_chat_on_cgmm);
+
+MODEM_CHAT_MATCHES_DEFINE(unsol_matches,
+			  MODEM_CHAT_MATCH("+CREG: ", ",", modem_cellular_chat_on_cxreg),
+			  MODEM_CHAT_MATCH("+CEREG: ", ",", modem_cellular_chat_on_cxreg),
+			  MODEM_CHAT_MATCH("+CGREG: ", ",", modem_cellular_chat_on_cxreg));
+
+MODEM_CHAT_MATCHES_DEFINE(abort_matches, MODEM_CHAT_MATCH("ERROR", "", NULL));
+
+MODEM_CHAT_MATCHES_DEFINE(connect_abort_matches,
+			  MODEM_CHAT_MATCH("ERROR", "", NULL),
+			  MODEM_CHAT_MATCH("BUSY", "", NULL),
+			  MODEM_CHAT_MATCH("NO ANSWER", "", NULL),
+			  MODEM_CHAT_MATCH("NO CARRIER", "", NULL),
+			  MODEM_CHAT_MATCH("NO DIALTONE", "", NULL));
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(init_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_MULT("AT+CGACT=0,1", allow_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_MULT("AT+CGACT=0,2", allow_match),
```

Comment:
For the HL7800 `AT+CGACT=0` will disable all instead of doing just 1 and 2 explicitly. Does that work for your modem?

Reply:
I found it was not needed for the Quectel modem, so I removed it instead :)